### PR TITLE
[SPARK-6428][SQL] Added explicit type for all public methods in sql/core

### PR DESCRIPTION
**Pull request by rxin**
Also implemented equals/hashCode when they are missing.

This is done in order to enable automatic public method type checking.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala [hunk 1]
```diff
@@ -91,19 +93,19 @@ case class HashOuterJoin(
       } else {
         joinedRow.withRight(rightNullRow).copy :: Nil
       }
-    )
+    }
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala]: This pair of braces is redundant.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala [hunk 2]
```diff
@@ -111,7 +113,7 @@ case class HashOuterJoin(
       } else {
         joinedRow.withLeft(leftNullRow).copy :: Nil
       }
-    )
+    }
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala]: Same as above.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 3]
```diff
@@ -198,6 +198,23 @@ private[sql] case class ParquetRelation2(
     case _ => false
   }
 
+  override def hashCode(): Int = {
+    if (shouldMergeSchemas) {
+      com.google.common.base.Objects.hashCode(
+        shouldMergeSchemas : java.lang.Boolean,
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Remove space before `:`?

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 4]
```diff
@@ -198,6 +198,23 @@ private[sql] case class ParquetRelation2(
...
+        paths.toSet,
+        maybeMetastoreSchema,
+        maybePartitionSpec)
+    } else {
+      com.google.common.base.Objects.hashCode(
+        shouldMergeSchemas : java.lang.Boolean,
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Same as above.

**Comment by liancheng** [comment]
Left some minor comments, otherwise LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala (hunk 2 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala]: but it is much more clear to have them. this block is pretty long.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala]: but it is much more clear to have them. this block is pretty long.

**Comment by marmbrus** [comment]
LGTM

**Comment by marmbrus** [comment]
backported to branch-1.3

**Pull request closed without merging**